### PR TITLE
esp32/esp32_rmt: New RMT capabilities.

### PR DESCRIPTION
**Pull request by jonathanhogg**
Following on from discussion in mattytrentini#11, I've picked up @mattytrentini's new RMT capabilities: new, better `tx_carrier` specification; `idle_level` setting; and more flexible ways of specifying pulses in `write_pulses()`.

Following on a "fixme" question in the working code, I implemented the ability to disable idle output (`config.tx_config.idle_output_en = 0`) by supplying `idle_level=None` when constructing the `RMT` object. However, I am unable to actually see this do anything useful when testing on real hardware with an oscilloscope.

I had naively assumed that it would leave the output high impedance when not transmitting, but as far as I can see the output still sits at 0V when idle. Trying with a TX carrier enabled also shows that the carrier continues to be generated when idle as well, so it doesn't control that. What is this setting actually for? The ESP-IDF documentation says nothing on the matter other than `idle_output_en` "[enables] the RMT output if idle". You got any experience of this @mattytrentini? If nobody knows what it's meant to do, I may take this logic back out again.

I have run a load of tests on an ESP32 board and everything else here seems to work as I would expect. Documentation has been updated to match the new logic.

Also, I have a question for @dpgeorge on error checking: following @mattytrentini's lead, I've put in fairly extensive parameter checking with useful error messages. However, is it preferable to be a bit simpler and more terse in order to keep the binary size down? I've seen other files where all the errors are just something like "parameter error".

### CODE REVIEW

**Comment by jonathanhogg** [comment]
I've taken the `idle_level=None` support out. I can't make it do anything useful for me with actual hardware, so it seems safer to leave it until someone else can explain it and a compelling reason to have it back in.

File: ports/esp32/esp32_rmt.c
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:99]: If you use `mp_obj_get_array_fixed_n(tx_carrien_obj, 3, &items)` then it'll automatically give a nice error message, you won't need to raise any yourself.  This whole first section of the if-statement can be replaced by that call.

File: ports/esp32/esp32_rmt.c
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:103]: No need to check for `mp_obj_is_int()` above, this function will raise an exception if it's not.

**Comment by dpgeorge** [comment]
> I've put in fairly extensive parameter checking with useful error messages. However, is it preferable to be a bit simpler and more terse in order to keep the binary size down?

In general, yes, better to be a bit terse to save flash space.  But at the same time be useful!  It's a hard balance, and in the core there are 3 levels of exception message detail (terse, normal, detailed).  But ports usually don't provide such configurability for error message detail.  In the end we don't want a port to overflow its flash because it's got too many long error messages...

**Comment by jonathanhogg** [comment]
Cool. I'll give this a bit of a tidy-up.

**Review by mattytrentini — APPROVED** [review]
Thanks for running with this @jonathanhogg! It looks in good shape. We probably need better examples of how to use the different methods of `write_pulses()` but that shouldn't hold up this merge, they can come later. Nice work!

File: ports/esp32/esp32_rmt.c
Comment (mattytrentini) [inline ports/esp32/esp32_rmt.c:140]: Should we display the idle level and, if configured, the carrier details?

Comment (jonathanhogg) [reply ports/esp32/esp32_rmt.c:140]: Yeah, I just copied in your simplification here. I wasn't sure if I should extend this to be complete or just save the bytes. Away for a few days, but I can fill this out when I get back if it seems like a good idea.

**Comment by jonathanhogg** [comment]
I have rebased this to mainline, fixed the documentation conflict and squashed to 2 commits.

I also added display of the `idle_level` to the print function – as suggested by @mattytrentini. The ESP IDF doesn't seem to provide a function for reading back the TX carrier configuration, so the only way to display this in the print function is to save these details to the Python object in the initialiser. This doesn't feel worth wasting the bytes on to me, but I'm happy to do this if necessary.

(OMG, I hate git rebasing so much.)

**Comment by mattytrentini** [comment]
@dpgeorge is there any chance we can get this in for the 1.13 release? It seems in good shape and should be low risk (only affects the ESP32 port and RMT is known to be in development). It adds some very useful features...

**Comment by jonathanhogg** [comment]
@dpgeorge should I continue to support this PR? If you're thinking that the RMT design needs something more substantial done to it, then I'm happy to work on that.

I'm using this API in production devices at the moment so I need to make some kind of call on whether that is a good idea if the features as designed aren't likely to make it to mainline.

File: docs/library/esp32.rst
Comment (dpgeorge) [inline docs/library/esp32.rst:193]: `None` is not supported anymore

**Review by dpgeorge — CHANGES_REQUESTED** [review]

**Comment by dpgeorge** [comment]
This looks like a good improvement, gives more flexibility and control over the RMT module.

It is a breaking change to the API... at least the constructor will raise an exception when used the old way, because the kwargs have changed name.

For the `write_pulses()` method: it's been changed to have 3 modes (per the comments in the code):
```python
# Mode 1: Constant duration
duration = 10000
data = (1,0,1,1,1,0,1,0,1)

# Mode 2: Always toggle (current implementation)
duration = (8000,11000,8000,11000,6000,13000,6000,3000,8000)
start = 1

# Mode 3: Most flexible
duration = (400,200,100,300,200,400)
data = (1,0,1,1,0,1)
```
These 3 modes are distinguished by the args being tuples/lists or integers, and also the `start` kwarg.  The `write_pulses()` method has 3 args but only 2 are ever used, so IMO it'd be cleaner to just make it take 2 positional (and unnamed) args and have the modes distinguished as:
```python
write_pulses(<int>, <tuple)  # mode 1
write_pulses(<tuple>, <int>)  # mode 2
write_pulses(<tuple>, <tuple>)  # mode 3
```
(tuple could also be a list)

I also think it's worth considering just having 3 separate method name, to make it less confusing, eg:
```python
write_pulses_const(duration, pulse_array)
write_pulses_toggle(duration_array, start_value)
write_pulses(duration_array, value_array)
```

**Comment by jonathanhogg** [comment]
Ta for the notes! I'll take a look at all of these alongside debugging #7403.

**Comment by jonathanhogg** [comment]
> These 3 modes are distinguished by the args being tuples/lists or integers, and also the `start` kwarg. The `write_pulses()` method has 3 args but only 2 are ever used, so IMO it'd be cleaner to just make it take 2 positional (and unnamed) args and have the modes distinguished as:
> 
> ```python
> write_pulses(<int>, <tuple)  # mode 1
> write_pulses(<tuple>, <int>)  # mode 2
> write_pulses(<tuple>, <tuple>)  # mode 3
> ```
> 
> (tuple could also be a list)

This is a great idea. I've gone for a slight variation of this – so that the previous `write_pulses(durations)` will still work – by making the second argument optional. So the API is:

```python
write_pulses(duration, data=1, /)
```

Where the three modes of operation are:

- `duration` is a tuple/list and `data` is an integer (or not given): use data as start output level and toggle
- `duration` is an integer and `data` is a tuple/list: use fixed duration and sequence of output levels
- `duration` and `data` are both tuple/list: specify durations and output levels explicitly

Anything else is an error. Arguably the default for `data` should be `not idle_level`.

I've also made a change to the `wait_done()` API to make the `timeout` keyword argument in milliseconds. It was documented as being in either 80Mhz clock ticks (in code) or RMT channel ticks (in documentation), but was actually in FreeRTOS timer ticks by behaviour. Switching to milliseconds feels more comprehensible.

I did a bunch of changes to the docs to try and make them read better and be more consistent in wording.

I've compressed into two commits and rebased on top of #7453 so this can be merged after or instead of that.

File: docs/library/esp32.rst
Comment (dpgeorge) [reply docs/library/esp32.rst:234]: > From inspecting the code I see that the acceptable values are "zero" or "non-zero"

Or maybe not... it could be odd/even.

File: docs/library/esp32.rst
Comment (jonathanhogg) [reply docs/library/esp32.rst:228]: Yes, good point. Will fix this.

File: docs/library/esp32.rst
Comment (jonathanhogg) [reply docs/library/esp32.rst:234]: Yes, more precisely it is even/odd. I thought perhaps to keep the docs as 0/1 as that's the logical value and the even/odd thing is an implementation detail that may change – I mean, the `++` to toggle is a bit of a hack.

File: ports/esp32/esp32_rmt.c
Comment (jonathanhogg) [reply ports/esp32/esp32_rmt.c:119]: I changed all of my original variables here to be unsigned `mp_uint_t` type as I noted that this change had been made in the mainline code.

File: ports/esp32/esp32_rmt.c
Comment (jonathanhogg) [reply ports/esp32/esp32_rmt.c:65]: Perhaps that is the best approach to all of the 0/1 inputs then?

File: ports/esp32/esp32_rmt.c
Comment (jonathanhogg) [reply ports/esp32/esp32_rmt.c:99]: Think I'll just make this switch then. I also notice that I've left these as `mp_int_t` when they can't be negative.

File: ports/esp32/esp32_rmt.c
Comment (jonathanhogg) [reply ports/esp32/esp32_rmt.c:140]: There's no API for retrieving the TX carrier state once you've set it and I wasn't convinced it was worth wasting the bytes in the object to store it just so it can be shown in the print function. However, I'm happy to do this if you think so.

Comment (dpgeorge) [reply ports/esp32/esp32_rmt.c:140]: Ah ok.  Just leave it as-is for now (ie don't print the TX carrier state).  Can be added later if it's really needed.

File: docs/library/esp32.rst
Comment (dpgeorge) [reply docs/library/esp32.rst:234]: Up to you what to do here.  I would lean towards using `mp_obj_is_true()` to convert the incoming Python values to 0/1.  This follows how pins work when doing `pin.value(x)` (x is converted to 0/1 using `mp_obj_is_true`).

**Comment by jonathanhogg** [comment]
OK, I think I've sorted all of your notes with this latest version. I've had to switch my logic to checking if `data` is not an array (tuple or list) instead of checking if it's an integer for mode 1. Logically an array can be converted to a boolean, so I'm slightly breaking the conceptual model here, but whatever.

**Comment by dpgeorge** [comment]
Thanks for updating, looks really good now.  Merged in 18e48a71ee69557a5340c8652f2e73e586063be3 and 0b3332c8e14db8a229070224955062f143b138cc

**Pull request closed without merging**
